File: 01-microvms/node/worker.js
'use strict';
// ============================================================================
// A real, persistent Node.js interpreter holding one session's state.
//
// The Node counterpart of worker.py, deliberately structured the same way so
// the two runtimes can be read side by side. Its context object survives
// suspend and resume because AWS checkpoints the VM's memory -- there is no
// save path, no serialization and no replay. When a resumed session still
// knows `balance`, it is the same process that set it.
//
// The VM is the security boundary, not this interpreter. Submitted code runs
// through vm.runInContext with full access to the context, which is acceptable
// only because the MicroVM around it is isolated and holds no AWS credentials.
// ============================================================================
const vm = require('vm');
const fs = require('fs');
const crypto = require('crypto');
const readline = require('readline');

// Total captured output per cell, matching the Python worker's cap. Keeps a
// runaway loop from returning a response too large for API Gateway.
const OUTPUT_LIMIT = 16000;

function main() {
  const started = process.hrtime.bigint();

  // Real initialization work, performed before the readiness line is printed,
  // so the image snapshot captures it and launches skip it entirely.
  const sales = [];
  for (let i = 0; i < 200000; i++) {
    sales.push([i % 12, ((i * 7919) % 10000) / 100]);
  }
  const totals = {};
  for (let m = 0; m < 12; m++) totals[m] = 0;
  for (const [m, v] of sales) totals[m] += v;

  // The persistent scope. Same role as the Python worker's namespace dict:
  // assignments made by submitted code land here and outlive the cell.
  const context = vm.createContext({ sales, totals, fs });

  const digest = crypto.createHash('sha256')
    .update(JSON.stringify(totals)).digest('hex');

  send({
    ready: true,
    rows: sales.length,
    init_ms: Number(process.hrtime.bigint() - started) / 1e6,
    dataset_sha256: digest,
    pid: process.pid,
  });

  // One JSON object per line in each direction, matching the Python worker so
  // the server implementations stay interchangeable.
  readline.createInterface({ input: process.stdin }).on('line', (line) => {
    let payload;
    try {
      payload = JSON.parse(line);
    } catch (error) {
      // Answer anyway. A silent iteration would hang the server waiting on a
      // response that never arrives.
      send({ ok: false, stdout: String(error) });
      return;
    }
    send(run(context, payload.code));
  });
}

function run(context, code) {
  let output = '';
  // Submitted code must never write to the real stdout: that is the protocol
  // channel back to the server, and a console.log would corrupt the stream.
  // Give the context its own console instead of redirecting the process.
  const capture = (...args) => {
    if (output.length < OUTPUT_LIMIT) {
      const text = args.map(a => typeof a === 'string' ? a : format(a)).join(' ');
      output += text.slice(0, OUTPUT_LIMIT - output.length) + '\n';
    }
  };
  context.console = { log: capture, error: capture, warn: capture, info: capture };

  const started = process.hrtime.bigint();
  let ok = true;
  try {
    vm.runInContext(code, context, { filename: '<cell>' });
  } catch (error) {
    ok = false;
    output += (error && error.stack ? error.stack : String(error)) + '\n';
  }
  return {
    ok,
    stdout: output,
    execution_ms: Number(process.hrtime.bigint() - started) / 1e6,
  };
}

// console.log-like rendering, so printing an object shows its contents rather
// than [object Object].
function format(value) {
  try {
    return require('util').inspect(value, { depth: 3, breakLength: 100 });
  } catch (error) {
    return String(value);
  }
}

function send(value) {
  process.stdout.write(JSON.stringify(value) + '\n');
}

main();

File: 01-microvms/python/worker.py
"""A real, persistent Python interpreter holding one session's state.

This process is the thing the whole demo is about. Its globals dict survives
suspend and resume because AWS checkpoints the VM's memory, not because anything
here is serialized -- there is no save path, no pickling and no replay. When a
resumed session still knows `balance`, it is the same process that set it.

The VM is the security boundary, not this interpreter. Submitted code runs through
exec with full access to the process, and that is acceptable only because the
MicroVM around it is isolated and holds no AWS credentials.
"""
import contextlib
import io
import json
import os
from pathlib import Path
import sys
import time
import traceback


class LimitedOutput(io.StringIO):
    """A StringIO that silently caps total captured output.

    Reports the full length back to the caller so a runaway print loop still
    completes at native speed, rather than blocking on a full pipe or returning
    a response too large for API Gateway.
    """

    def write(self, value):
        remaining = 16000 - self.tell()
        if remaining > 0:
            super().write(value[:remaining])
        return len(value)


def main():
    """Announce readiness, then serve cells from stdin.

    Anything done before the readiness line is printed becomes part of the image
    snapshot, because the /ready hook does not pass until this finishes. There
    is deliberately nothing here to preload -- but this is where it would go.

    The protocol is one JSON object per line in each direction. Line-delimited
    JSON over pipes avoids any dependency and keeps the interpreter isolated in
    its own process, so killing a hung cell cannot take down the HTTP server.
    """
    # The session namespace. Assignments made by submitted code land here and
    # outlive the cell; Path is seeded so a cell can touch the filesystem
    # without an import.
    namespace = {"__name__": "__session__", "Path": Path}

    # Nothing is preloaded, so this returns immediately -- but it still has to
    # be printed, because the server blocks on it and only then does /ready
    # pass. That handshake is what decides when the snapshot is taken.
    print(json.dumps({"ready": True, "pid": os.getpid()}), flush=True)

    for line in sys.stdin:
        try:
            payload = json.loads(line)
            output = LimitedOutput()
            started = time.perf_counter()
            # Redirect both streams so a traceback is returned as data instead
            # of vanishing into the container log.
            with contextlib.redirect_stdout(output), contextlib.redirect_stderr(output):
                try:
                    # Same dict for globals and locals, so assignments persist
                    # into the session rather than dying with the exec frame.
                    exec(compile(payload["code"], "<cell>", "exec"), namespace, namespace)
                    ok = True
                except BaseException:
                    # BaseException, not Exception: SystemExit from submitted
                    # code calling sys.exit() must be reported, not kill the session.
                    traceback.print_exc(limit=3)
                    ok = False
            print(json.dumps({"ok": ok, "stdout": output.getvalue(),
                              "execution_ms": round((time.perf_counter() - started) * 1000, 2)}), flush=True)
        except Exception as exc:
            # Malformed request line. Answer anyway; a silent loop iteration
            # would hang the server waiting on a response that never comes.
            print(json.dumps({"ok": False, "stdout": str(exc)}), flush=True)


if __name__ == "__main__":
    main()
